pvsystemprofiler/tilt_azimuth_study.py

```diff
@@ -0,0 +1,281 @@
+    def find_fit_costheta(self):
+        data = np.max(self.data_matrix, axis=0)
+        s1 = cvx.Variable(len(data))
+        s2 = cvx.Variable(len(data))
+        cost = 1e1 * cvx.norm(cvx.diff(s1, k=2), p=2) + cvx.norm(s2[self.clear_index])
+        objective = cvx.Minimize(cost)
+        constraints = [
+            data == s1 + s2,
+            s1[365:] == s1[:-365]
+        ]
+        problem = cvx.Problem(objective, constraints)
+        problem.solve(solver='MOSEK')
+        self.scale_factor_costheta = s1.value
+        self.costheta_fit = self.data_matrix / np.max(s1.value)
+        return
+
+    def ground_truth_costheta(self):
+        phi_true_value_2d = np.tile(np.deg2rad(self.phi_true_value),
+                                    (self.daily_meas, self.num_days))
+        beta_true_value_2d = np.tile(np.deg2rad(self.beta_true_value),
+                                     (self.daily_meas, self.num_days))
+        gamma_true_value_2d = np.tile(np.deg2rad(self.gamma_true_value),
+                                      (self.daily_meas, self.num_days))
+        X = np.array([self.omega, self.delta, phi_true_value_2d])
+        self.costheta_ground_truth = \
+            self.func(X, beta_true_value_2d, gamma_true_value_2d)
+        return
+
+    def estimate_costheta(self):
+        phi_estimate_2d = np.tile(np.deg2rad(self.latitude_estimate), (self.daily_meas, self.num_days))
+        beta_estimate_2d = np.tile(np.deg2rad(self.tilt_estimate), (self.daily_meas, self.num_days))
+        gamma_estimate_2d = np.tile(np.deg2rad(self.azimuth_estimate), (self.daily_meas, self.num_days))
+        X = np.array([self.omega, self.delta, phi_estimate_2d])
+        self.costheta_estimated = self.func(X, beta_estimate_2d, gamma_estimate_2d)
+        return
+
+    def find_daytime_threshold_quantile_seasonality(self):
+        m = cvx.Parameter(nonneg=True, value=10 ** 6)
+        t = cvx.Parameter(nonneg=True, value=0.9)
+        y = np.quantile(self.data_matrix, .9, axis=0)
+        x1 = cvx.Variable(len(y))
+        x2 = cvx.Variable(len(y))
+        if self.data_matrix.shape[1] > 365:
+            constraints = [
+                x2[365:] == x2[:-365], x1 + x2 == y
+            ]
+        else:
+            constraints = [x1 + x2 == y]
+        c1 = cvx.sum(1 / 2 * cvx.abs(x1) + (t - 1 / 2) * x1)
+        c2 = cvx.sum_squares(cvx.diff(x2, 2))
+        objective = cvx.Minimize(c1 + m * c2)
+        prob = cvx.Problem(objective, constraints=constraints)
+        prob.solve(solver='MOSEK')
+        return x2.value
+
+    def select_days(self):
+        if self.scsf:
+            self.boolean_daytime_range = self.boolean_daytime * self.day_range
+        else:
+            self.boolean_daytime_range = self.boolean_daytime * self.clear_index * self.day_range
+            self.delta_f = self.delta[self.boolean_daytime_range]
+            self.omega_f = self.omega[self.boolean_daytime_range]
+
+    def run_curve_fit(self, bootstrap_iterations=None):
```

Comment:
The core algorithm—curve fitting the geometric equation—needs to be refactored to a stand-along file in `pvsystemprofiler/algorithms`. This one gets tricky with the code organization and naming because we're using one equation/algorithm that _can_ be used to estimate lat+tilt+az or any subset of those parameters. Do this last, and ping me when you're ready to do this refactoring.